runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/testing/TestDataflowRunner.java

```diff
@@ -118,14 +120,16 @@ DataflowPipelineJob run(Pipeline pipeline, DataflowRunner runner) {
             try {
               for (;;) {
                 Optional<Boolean> result = checkForSuccess(job);
-                if (result.isPresent()) {
+                if (result.isPresent() && (!result.get() || atMaxWatermark(job))) {
                   return result;
                 }
                 Thread.sleep(10000L);
               }
             } finally {
-              LOG.info("Cancelling Dataflow job {}", job.getJobId());
-              job.cancel();
+              if (!job.getState().isTerminal()) {
```

Comment:
The reason I add this check is because job can be canceled from UI by user or timeout (waitUntilFinish() in above few lines) first. Then this thread will be aware and cancel job again which will cause exception and terminate test run.